Filtering with FilterDuplicates alters precision

**Describe the bug**
<!-- A clear and concise description of what the bug is. -->

Rounding is done in-place with FilterDuplicates, which causes information loss and means potentially fitting to possibly incorrect data.

**To Reproduce**
<!-- Steps to reproduce the behavior. A minimal reproducing set of python commands is ideal.

If the problem involves a specific molecule or file, please upload that as well. -->
```
df = pd.read_csv("dataset.csv")
filtered = CurationWorkflow.apply(
    df,
    CurationWorkflowSchema(
        component_schemas=[
            filtering.FilterDuplicatesSchema(
                mole_fraction_precision=2,
            ),
        ]
    )
)
assert list(df["Mole Fraction 1"]) == list(filtered["Mole Fraction 1"])
```
**Output**
<!-- The full error message (may be large, that's fine. Better to paste too much than too little.) -->

**Computing environment (please complete the following information):**
 - Operating system
 - Output of running `conda list`

**Additional context**
<!-- Add any other context about the problem here. -->

[dataset.csv](https://github.com/user-attachments/files/19280404/dataset.csv)

Unless I'm missing something obvious, you already have a MWE in place?

Tidied up a tiny bit for lazier copy-paste:
```python
In [1]: from openff.evaluator.datasets.curation.components import filtering
   ...: from openff.evaluator.datasets.curation.workflow import (
   ...:     CurationWorkflowSchema,
   ...:     CurationWorkflow,
   ...: )
   ...: import pandas as pd
   ...:
   ...: df = pd.read_csv("dataset.csv")
   ...: filtered = CurationWorkflow.apply(
   ...:     df,
   ...:     CurationWorkflowSchema(
   ...:         component_schemas=[
   ...:             filtering.FilterDuplicatesSchema(
   ...:                 mole_fraction_precision=2,
   ...:             ),
   ...:         ]
   ...:     ),
   ...: )
   ...: assert list(df["Mole Fraction 1"]) == list(filtered["Mole Fraction 1"]), (
   ...:     df["Mole Fraction 1"],
   ...:     filtered["Mole Fraction 1"],
   ...: )
   ...:
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
Cell In[1], line 19
      8 df = pd.read_csv("dataset.csv")
      9 filtered = CurationWorkflow.apply(
     10     df,
     11     CurationWorkflowSchema(
   (...)
     17     ),
     18 )
---> 19 assert list(df["Mole Fraction 1"]) == list(filtered["Mole Fraction 1"]), (
     20     df["Mole Fraction 1"],
     21     filtered["Mole Fraction 1"],
     22 )

AssertionError: (0    0.7912
1    0.2112
Name: Mole Fraction 1, dtype: float64, 0    0.79
1    0.21
Name: Mole Fraction 1, dtype: float64)

--- a/openff/evaluator/_tests/test_datasets/test_curation/test_filtering.py
+++ b/openff/evaluator/_tests/test_datasets/test_curation/test_filtering.py
@@ -43,6 +43,10 @@ from openff.evaluator.datasets.curation.components.filtering import (
     FilterDuplicates,
     FilterDuplicatesSchema,
 )
+from openff.evaluator.datasets.curation.workflow import (
+    CurationWorkflow,
+    CurationWorkflowSchema,
+)
 from openff.evaluator.datasets.utilities import data_frame_to_substances
 from openff.evaluator.properties import Density, EnthalpyOfMixing
 from openff.evaluator.substances import Component, MoleFraction, Substance
@@ -1153,3 +1157,50 @@ def test_filter_by_environment_per_component():
     assert {*pure_data["Component 1"].unique()} == {"O", "C"}
 
     assert len(binary_data) == 2
+
+
+def test_curation_does_not_alter_precision():
+    """See issue #629"""
+    # most faithful in-memory representation of how pandas sees missing data in CSVs
+    nan = numpy.nan
+
+    # copy-pasted from dataset.csv in linked issue
+    data_frame = pandas.DataFrame.from_dict(
+        {
+            "Unnamed: 0": {0: 267, 1: 268},
+            "Id": {
+                0: "4031d37fd75649b7b143b42e62ca3338",
+                1: "eeaf958a3e8d4ee1931bfb8a5258ef49",
+            },
+            "Temperature (K)": {0: 298.15, 1: 298.15},
+            "Pressure (kPa)": {0: 101.0, 1: 101.0},
+            "Phase": {0: "Liquid", 1: "Liquid"},
+            "N Components": {0: 2, 1: 2},
+            "Component 1": {0: "NCCO", 1: "NCCO"},
+            "Role 1": {0: "Solvent", 1: "Solvent"},
+            "Mole Fraction 1": {0: 0.7912, 1: 0.2112},
+            "Density Value (g / ml)": {0: 1.002592, 1: 0.9825660000000004},
+            "Density Uncertainty (g / ml)": {0: 0.000101, 1: 0.0001015},
+            "Source": {0: "10.1021/je400184t", 1: "10.1021/je400184t"},
+            "Component 2": {0: "c1ccncc1", 1: "c1ccncc1"},
+            "Role 2": {0: "Solvent", 1: "Solvent"},
+            "Mole Fraction 2": {0: 0.2088, 1: 0.7888},
+            "EnthalpyOfMixing Value (kJ / mol)": {0: nan, 1: nan},
+            "EnthalpyOfMixing Uncertainty (kJ / mol)": {0: nan, 1: nan},
+            "Exact Amount 1": {0: nan, 1: nan},
+            "Exact Amount 2": {0: nan, 1: nan},
+        }
+    )
+
+    filtered = CurationWorkflow.apply(
+        data_frame,
+        CurationWorkflowSchema(
+            component_schemas=[
+                FilterDuplicatesSchema(
+                    mole_fraction_precision=2,
+                ),
+            ]
+        ),
+    )
+
+    assert list(data_frame["Mole Fraction 1"]) == list(filtered["Mole Fraction 1"])

--- a/openff/evaluator/datasets/curation/components/filtering.py
+++ b/openff/evaluator/datasets/curation/components/filtering.py
@@ -140,9 +140,7 @@ class FilterDuplicates(CurationComponent):
             filtered_data.append(sorted_filtered_data)
 
         filtered_data = pandas.concat(filtered_data, ignore_index=True, sort=False)
-
-        original_filtered_data = data_frame[data_frame["Id"].isin(filtered_data["Id"])]
-        return original_filtered_data
+        return filtered_data
 
 
 class FilterByTemperatureSchema(CurationComponentSchema):

--- a/openff/evaluator/datasets/curation/components/filtering.py
+++ b/openff/evaluator/datasets/curation/components/filtering.py
@@ -140,7 +140,9 @@ class FilterDuplicates(CurationComponent):
             filtered_data.append(sorted_filtered_data)
 
         filtered_data = pandas.concat(filtered_data, ignore_index=True, sort=False)
-        return filtered_data
+
+        original_filtered_data = data_frame[data_frame["Id"].isin(filtered_data["Id"])]
+        return original_filtered_data
 
 
 class FilterByTemperatureSchema(CurationComponentSchema):

